src/data_registry/data_registry.cpp

```diff
@@ -0,0 +1,68 @@
+
+#include "data_registry.hpp"
+#include "registry_entries.hpp"
+
+data_registry::data_registry() {
+
+    // Initialise sensor data entry
+    sensor_data* sd = new sensor_data();
+    registry[SENSOR_DATA_ID].id = SENSOR_DATA_ID;
+    registry[SENSOR_DATA_ID].data = (void*) sd;
+
+}
+
+data_registry::~data_registry() {
+
+    global_lock.lock();
+    
+    for (auto& pair : registry) {
+        pair.second.lock.lock();
+        delete pair.second.data;
+        pair.second.lock.unlock();
+    }
+
+    global_lock.unlock();
+}
+
+data_registry& data_registry::get_instance() {
+    static data_registry instance;
+    return instance;
+}
+
+void* data_registry::acquire_data(const std::string& id) {
+
+    void* data;
+    global_lock.lock();
+
+    // Check if entry exists
+    if (!registry.count(id)) {
+        // TODO: Throw exception or print to IO-thread?
+        global_lock.unlock();
+        return nullptr;
+    }
+
+    registry_entry& entry = registry[id];
+    entry.lock.lock();
+    data = entry.data;
+
+    entry.lock.unlock();
```

Comment:
This shouldn't be unlocked... But removing this unlock will always cause a deadlock with global_lock if acquire is called again on this object by another thread before a release was called. Maybe global_lock is unnecessary since the data it locks is always only read (entries in the map) and never written to.